src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/markduplicates/MarkDuplicates.java

```diff
@@ -0,0 +1,58 @@
+package org.broadinstitute.hellbender.tools.dataflow.transforms.markduplicates;
+
+import com.google.cloud.dataflow.sdk.transforms.*;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PCollectionList;
+import com.google.cloud.dataflow.sdk.values.PCollectionView;
+import htsjdk.samtools.SAMFileHeader;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+/**
+ * The main transform for MarkDuplicates. Takes reads, returns reads with some reads marked as duplicates.
+ */
+public class MarkDuplicates extends PTransform<PCollection<GATKRead>, PCollection<GATKRead>> {
+    private static final long serialVersionUID = 1l;
+    
+    //private vs non-primary alignments
+    private enum ReadsPartition {
+        PRIMARY, NOT_PRIMARY
+    }
+
+    /**
+     * The header for all reads processed.
+     */
+    private final PCollectionView<SAMFileHeader> header;
+
+    public MarkDuplicates(final PCollectionView<SAMFileHeader> header) {
+        this.header = header;
+    }
+
+    @Override
+    public PCollection<GATKRead> apply(final PCollection<GATKRead> preads) {
+        final PCollectionList<GATKRead> readsPartitioned = partitionReadsByPrimaryNonPrimaryAlignment(preads);
+
+        final PCollection<GATKRead> fragments = readsPartitioned.get(ReadsPartition.PRIMARY.ordinal());
```

Comment:
My suspicion would match Louis'. Remember, the elements are assumed to be immutable. In that context doing copies is wasteful work. Even if the code did copies today (I don't think it does), those copies may disappear in the next version.

I also suspect that the current code doesn't care if the early mutations are visible in the later code.